### ISSUE TITLE
Unpack failure when a filename contains single quotes

### ISSUE THREAD

**Issue by NicolasToussaint** [fossology/fossology#1397]
### Description

Unpack failure when a filename contains the character `'`.

#### How to reproduce

1. Upload a package containing a file with a quote in its name (eg. `foo'bar.baz`)
1. The Unpack stage will fail with following logs:
```
2019-07-09 12:49:12 ununpack [0] :: JOB[102].ununpack[1621.localhost]: "sh: 1: Syntax error: Unterminated quoted string"
2019-07-09 12:49:12 ununpack [0] :: JOB[102].ununpack[1621.localhost]: "FATAL utils.c.1585: Unable to calculate SHA256 of /srv/fossology/repository/6764497f58c4/ununpack/469540l/94c2650a0e7a9b7b7ad839860ccb1d9231232b4c.7475b9a3872350af0d2c2300446b0445.240.dir/upload_01.dir/upload_01/foo'bar.baz"
2019-07-09 12:49:12 ununpack [0] :: JOB[102].ununpack[1621.localhost]: agent failed with error code 56
2019-07-09 12:49:12 ununpack [0] :: JOB[102].ununpack[1621.localhost]: agent failed, code: 56
```

The problem come from this line, as the filename is included unprotected:
https://github.com/fossology/fossology/blob/a26c0cd8ec0dca5063248667d734f86cc5c23f08/src/ununpack/agent/utils.c#L1576

Replacing single quotes fixes with `"sha256sum \"%s\""` fixes the single-quote problem, but of course now it fails with filenames containing double quotes :)

I don't know if there is already an escape function available in the source code to reuse ; but I'll look for a quick workaround for now.

#### Screenshots

![image](https://user-images.githubusercontent.com/22956329/60890553-cad21500-a25b-11e9-8757-47d34975d2c6.png)

### Versions

* Last commit id on master: https://github.com/fossology/fossology/commit/53849883c7a95f699f8b2198a1849aa6926dee6f
* Operating System (lsb_release -a): Docker build

**Comment by GMishx** [comment]
Thanks for pointing this out @NicolasToussaint. There is an issue running `sha256sum` as a child process if the path contains any kind of special characters.
This was a quick fix to enclose the path in `'` to accommodate special characters like @, space, etc. But it will fail for another `'`.

The best way will be to escape such characters before constructing the path.

**Comment by NicolasToussaint** [comment]
I'm looking into this, and it turns out that simply escaping the single quotes contained in the file name does not work, to my surprise (at least in the Docker environment)
```
root@7dffae3fba39:~# sha256sum 'foo\'bar'                                                                                               
> 
```

So the only solution I found was to embed the filename in double quotes, and escape all necessary characters: `"`, `\` and `$` (did I miss any?)

I have a working fix, except that the filenames containing `\` characters still generate an error - though it doesn't look related to the SHA256SUM problem, but rather a postgres error:

From the Ununpack job logs:
```
2019-08-06 08:05:49 ununpack [0] :: JOB[216].ununpack[1629.localhost]: "ERROR: Invalid filename 'DA39A3EE5E6B4B0D3255BFEF95601890AFD80709.D41D8CD98F00B204E9800998ECF8427E.\e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b85'"
2019-08-06 08:05:49 ununpack [0] :: JOB[216].ununpack[1629.localhost]: "row number 0 is out of range 0..-1"
```

Further, testing with an older version, I realized that `\` characters were turned into `~`.

**Comment by fogninid** [comment]
have you tried using TaintString?: https://github.com/fossology/fossology/blob/a26c0cd8ec0dca5063248667d734f86cc5c23f08/src/ununpack/agent/utils.c#L175
Its description says that it's supposed to do exactly what you need.

__OR__ you could rewrite the `popen` using a proper `fork+exec` avoiding the shell and escaping problems entirely. For example see here for inspiration: https://stackoverflow.com/a/19669367/5162472

**Comment by NicolasToussaint** [comment]
Thanks @fogninid for the suggestions ;

I don't think that the `TaintString` answers the problem, since it only escapes `'`, `\` and `%` characters which is not our need here: 
- `%` is not a problem
-  `\` causes trouble down the road (I think)
- `'` cannot be included in single quote strings in bash (or dash)
  - See [man bash](https://linux.die.net/man/1/bash)

>Enclosing characters in single quotes preserves the literal value of each character within the quotes. A single quote may not occur between single quotes, even when preceded by a backslash. 


So, I suggest that either:
1. We modify the `TaintString` accordingly
2. We carry on with double quote string - and fix the `\` problem
3. We change the way the sha256sum is computed
  - either via `fork+exec` as you suggested
  - or via an internal algorithm written in C

**Comment by NicolasToussaint** [comment]
Now I think about it, why not embed the SHA256 calculation (like it was done for MD5 and SHA1) ?

I could include one of the following existing implementations:
- https://github.com/intel/isa-l_crypto/tree/master/sha256_mb
- https://github.com/amosnier/sha-2
- https://github.com/ckolivas/cgminer/blob/master/sha2.c#L93

What do you think ?

**Comment by fogninid** [comment]
You are absolutely right about not using `TaintString`: the %s replacement has no business running on user provided input. IMHO there are only 2 correct solutions to escaping: avoid it or don't write it yourself. 
I am quite sure that other places in the code use it in an unsafe way, can anybody also have a look into it and open another ticket (or must I open one, I don't have much time right now)? I would look closely into ExtractDisk, ExtractISO and ExtractAR. The best solution would be to replace any `system` or `popen` calls with safer `fork+exec` and avoid the shell (and escaping) completely, but it's probably a high effort endeavour. 

Embedding the SHA256 calculation is probably the best solution for the current issue: you avoid a fork and you read each file only once if you perform the computation in the `checksums` functions.
The trick is only to find a correct C implementation whose license is compatible with fossology, @mcjaeger can you review the implementations found by @NicolasToussaint?

**Comment by NicolasToussaint** [comment]
Actually, I do believe that embedding the sha256 calculation is the best way too, like it is done for other checksums.

About the references I proposed:
- https://github.com/intel/isa-l_crypto/tree/master/sha256_mb => not straight forward to integrate
- https://github.com/amosnier/sha-2 => no support for multipart computation
- https://github.com/ckolivas/cgminer/blob/master/sha2.c#L93 => Looks just right, but...

When I build with the source code from the library, I get PIC related errors (`relocation R_X86_64_32 against  can not be used when making a shared object; recompile with -fPIC`).

It builds successfully with the `-fPIC`option, but then I get weird PHP / UTF8 display error when loading the ununpack job page ...

So I'm trying to sort this out...
